### PR TITLE
feat(pinterest): add ldp support

### PR DESCRIPTION
**Pull request by Gauravudia**
## PR Description

Add Limited Data Processing (LDP) support

## Notion ticket

[Ticket link](https://www.notion.so/rudderstacks/Pinterest-LDP-Support-6fbe0cb473e74a6a9a5b7d5e500f8090)

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/810)

### CODE REVIEW

File: src/utils/utils.js [hunk 1]
```diff
@@ -639,6 +639,11 @@ function getDefinedTraits(message) {
       message?.context?.traits?.lastname ||
       message?.context?.traits?.last_name,
     name: message?.context?.traits?.name || message?.context?.traits?.Name,
+    state:
+      message?.context?.traits?.state ||
+      message?.context?.traits?.State ||
+      message?.context?.traits?.address?.state ||
+      message?.context?.traits?.address?.State,
```
Comment (ItsSudip) [inline src/utils/utils.js:646]: Looks like a breaking change. Have we checked other destinations which are using this function?

File: src/integrations/PinterestTag/browser.js [hunk 2]
```diff
@@ -229,13 +268,15 @@ export default class PinterestTag {
...
+    this.setLdp(rudderElement.message);
     window.pintrk("track", event, pageObject);
   }
 
   identify() {
     const email = this.analytics.userTraits && this.analytics.userTraits.email;
     if (email) {
-      window.pintrk("set", { em: email });
+      const ldpObject = this.generateLdpObject();
+      window.pintrk('set', { em: email, ...ldpObject });
```
Comment (ItsSudip) [inline src/integrations/PinterestTag/browser.js:279]: What happened if `ldpObject` is undefined or empty?

File: src/integrations/PinterestTag/browser.js [hunk 3]
```diff
@@ -73,6 +74,43 @@ export default class PinterestTag {
...
+      };
+    }
+
+    const optOutType = message.context.traits?.optOutType || message.properties?.optOutType;
+    const { state, country } = getDefinedTraits(message);
+    return {
+      opt_out_type: optOutType,
+      st: sha256(state || ''),
+      country: sha256(country || ''),
+    };
```
Comment (ItsSudip) [inline src/integrations/PinterestTag/browser.js:106]: If optOutType is undefined, do we need to set it?

File: src/integrations/PinterestTag/browser.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 /* eslint-disable class-methods-use-this */
+import sha256 from 'sha256';
```
Comment (ItsSudip) [inline src/integrations/PinterestTag/browser.js:2]: We have already used sha256 from `crypto-js`, can't we use that one here?

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: src/integrations/PinterestTag/browser.js (hunk 3 above)
Comment (Gauravudia) [reply src/integrations/PinterestTag/browser.js]: code updated. Setting `optOutType` to empty strings If it is undefined.

File: src/integrations/PinterestTag/browser.js (hunk 4 above)
Comment (Gauravudia) [reply src/integrations/PinterestTag/browser.js]: used sha256 from `crypto-js`

File: src/integrations/PinterestTag/browser.js (hunk 2 above)
Comment (Gauravudia) [reply src/integrations/PinterestTag/browser.js:277]: `generateLdpObject()` will always return non empty ldpObject

Comment (Gauravudia) [reply src/integrations/PinterestTag/browser.js:277]: generateLdpObject() will return non empty ldpObject

File: src/utils/utils.js (hunk 1 above)
Comment (Gauravudia) [reply src/utils/utils.js:646]: Checked for Clevertap.

```
rudderanalytics.identify("userId123", {
          email: "clevertap_test@email.com",
          phone: "8123099213",
          name: "Clevertap Test",
          state: "UP",
          country: "IN"
        });
```
Values are coming perfectly!
Screenshot 2022-12-26 at 1 58 15 PM

**Review by shrouti1507 — APPROVED** [review]

**Review by MoumitaM — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**